### ISSUE TITLE
anomaly on Hoare logic `match` when the precondition is an existential

### ISSUE THREAD

**Issue by fdupress** [EasyCrypt/easycrypt#565]
MFE

```
require import List.

module M = {
  var r : int option

  proc hd(xs : int list) = {
    r <- match r with | None => None | Some _ => None end;
    match xs with
    | [] => { }
    | x :: xs => r <- Some x;
    end;
    return r;
  }
}.

hoare toto &m y ys:
     M.hd: arg = y :: ys ==> res = Some y.
proof.
proc. sp.
match. (* Fails with "anomaly: EcLib.EcCoreGoal.InvalidGoalShape" *)
```

**Comment by fdupress** [comment]
This was introduced by #510 .

**Comment by fdupress** [comment]
And also triggers:
- when the precondition merely _contains_ an existential quantifier (and perhaps a universal quantifier?);
- in one-sided `match` applications in pRHL.

**Comment by strub** [comment]
Fix by #577
